cartographer_ros/cartographer_ros/metrics/internal/gauge.h

```diff
@@ -63,8 +58,11 @@ class Gauge : public ::cartographer::metrics::Gauge {
 
  private:
   void Add(const double value) {
-    double current = Value();
-    Set(current + value);
+    double expected = value_.load();
+    double new_value = expected + value;
+    while (!value_.compare_exchange_weak(expected, new_value)) {
+      new_value = expected + value;
```

Comment:
you need to load expected again, otherwise the loop can be endless

Reply:
I think compare_exchange_weak does the load when it fails.

Reply:
expected is updated by reference

https://en.cppreference.com/w/cpp/atomic/atomic/compare_exchange
```
Atomically compares the object representation of *this with that of expected, and if those are bitwise-equal, replaces the former with desired (performs read-modify-write operation). Otherwise, loads the actual value stored in *this into expected (performs load operation). 
```